src/tests/Microsoft.VisualStudio.Composition.Tests/ThreadSafetyTests.cs

```diff
@@ -142,50 +143,55 @@ public void LazyOfNonSharedPartConstructsOnlyOneInstanceAcrossThreadsV2(IContain
+                var root = container.GetExportedValue<PartThatImportsNonSharedPart>();
 
-            // We carefully orchestrate the scheduling here to make sure we have threads ready to go
-            // (not just queued to the threadpool) to maximize concurrent execution.
-            var ready = new CountdownEvent(2);
-            var go = new ManualResetEventSlim();
+                // We carefully orchestrate the scheduling here to make sure we have threads ready to go
+                // (not just queued to the threadpool) to maximize concurrent execution.
+                var ready = new CountdownEvent(2);
+                var go = new ManualResetEventSlim();
 
-            var t1 = Task.Run(delegate
-            {
-                ready.Signal();
-                ready.Wait();
-                return root.NonSharedPart.Value;
-            });
+                var t1 = Task.Run(delegate
+                {
+                    ready.Signal();
+                    ready.Wait();
+                    return root.NonSharedPart.Value;
+                });
 
-            var t2 = Task.Run(delegate
-            {
-                ready.Signal();
-                ready.Wait();
-                return root.NonSharedPart.Value;
-            });
+                var t2 = Task.Run(delegate
+                {
+                    ready.Signal();
+                    ready.Wait();
+                    return root.NonSharedPart.Value;
+                });
 
-            var unblockingTask = Task.Run(async delegate
-            {
-                // We artificially block the constructor of the non-shared part
-                // to maximize the window that can result in multiple instances being created.
-                ready.Wait();
-                await Task.Delay(TestUtilities.ExpectedTimeout);
+                var unblockingTask = Task.Run(async delegate
+                {
+                    // We artificially block the constructor of the non-shared part
+                    // to maximize the window that can result in multiple instances being created.
+                    ready.Wait();
+                    await Task.Delay(TestUtilities.ExpectedTimeout);
 
-                // Now unblock the other threads.
-                NonSharedPart.UnblockCtor.Set();
+                    // Now unblock the other threads.
+                    NonSharedPart.UnblockCtor.Set();
 
-                // Pri-2: verify that the constructor was only invoked once.
-                if (!permitMultipleInstancesOfNonSharedPart)
-                {
-                    Assert.Equal(1, NonSharedPart.CtorInvocationCounter);
-                }
-            });
+                    // Pri-2: verify that the constructor was only invoked once.
+                    if (!permitMultipleInstancesOfNonSharedPart)
+                    {
+                        t1.Wait();
+                        t2.Wait();
```

Comment:
Given what you found about test classes being what get parallelized rather than individual methods, I suspect that _this_ change where you added these Waits is probably what will fix it. I do see a narrow opportunity for a race condition which you're addressing here.
I would support you removing the lock object and just keeping the task.Wait calls here if you're interested, both because I think it will suffice, and because if it won't suffice, I'd want to know that too. :)